graph/cassandra/quadstore.go

```diff
@@ -0,0 +1,378 @@
+			continue
+		}
+		query := fmt.Sprint("UPDATE ", table, `
+		SET deleted = deleted + ?
+		WHERE
+		subject = ? AND
+		predicate = ? AND
+		object = ? AND
+		label = ?
+		`)
+		data.Query(query, []int64{d.ID.Int()}, q.Subject, q.Predicate, q.Object, q.Label)
+	}
+	for _, dir := range []quad.Direction{quad.Subject, quad.Predicate, quad.Object, quad.Label} {
+		if q.Get(dir) == "" {
+			continue
+		}
+		query := fmt.Sprint("UPDATE nodes SET ", dir, "_count = ", dir, "_count - 1 WHERE node = ?")
+		count.Query(query, q.Get(dir))
+	}
+}
+
+func (qs *QuadStore) addDeltaToLog(d graph.Delta, data *gocql.Batch, size int64) {
+	data.Query(
+		"INSERT INTO log (id, size, timestamp, action, subject, predicate, object, label) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
+		d.ID.Int(),
+		size,
+		d.Timestamp,
+		d.Action.String(),
+		d.Quad.Subject,
+		d.Quad.Predicate,
+		d.Quad.Object,
+		d.Quad.Label,
+	)
+}
+
+func (qs *QuadStore) updateMetadata(data *gocql.Batch, size int64, horizon int64) {
+	data.Query("UPDATE metadata SET value = ? WHERE meta_key = ?", size, "size")
+	data.Query("UPDATE metadata SET value = ? WHERE meta_key = ?", horizon, "horizon")
+}
+
+func (qs *QuadStore) checkValid(delta graph.Delta) bool {
+	var created []int64
+	var deleted []int64
+	q := delta.Quad
+	err := qs.sess.Query(
+		`SELECT created, deleted
+		FROM quads_by_s
+		WHERE
+		subject = ? AND
+		predicate = ? AND
+		object = ? AND
+		label = ?
+		`, q.Subject, q.Predicate, q.Object, q.Label).Scan(&created, &deleted)
+	if err != nil && err != gocql.ErrNotFound {
+		return false
+	}
+	if err == gocql.ErrNotFound {
+		if delta.Action == graph.Add {
+			return true
+		}
+		// Can't delete something that doesn't exist.
+		return false
+	}
+
+	// err == nil at this point
+
+	if len(created) == len(deleted) {
+		if delta.Action == graph.Add {
+			return true
+		}
+		return false
+	}
+	if delta.Action == graph.Add {
+		return false
+	}
+	return true
+}
+
+func (qs *QuadStore) ApplyDeltas(deltas []graph.Delta, ignoreOpts graph.IgnoreOpts) error {
+	// Precheck existence.
+	for _, d := range deltas {
+		if d.Action != graph.Add && d.Action != graph.Delete {
+			return errors.New("cassandra: invalid action")
+		}
+		switch d.Action {
+		case graph.Add:
+			if !qs.checkValid(d) {
+				if ignoreOpts.IgnoreDup {
+					continue
+				} else {
+					return graph.ErrQuadExists
+				}
+			}
+		case graph.Delete:
+			if !qs.checkValid(d) {
+				if ignoreOpts.IgnoreMissing {
+					continue
+				} else {
+					return graph.ErrQuadNotExist
+				}
+			}
+		}
+	}
+	// Write the data.
+	batch := qs.sess.NewBatch(gocql.LoggedBatch)
+	batch.Cons = qs.writeConsistency
+	counterBatch := qs.sess.NewBatch(gocql.CounterBatch)
+	counterBatch.Cons = qs.writeConsistency
+	newSize := qs.size
+	newHorizon := qs.horizon
+	for _, d := range deltas {
+		if d.Action == graph.Add {
+			newSize++
+			qs.addDeltaToLog(d, batch, newSize)
+			qs.addQuadToBatch(d, batch, counterBatch)
```

Comment:
The batching here is somewhat dangerous from a performance perspective. The issue is that you're batching multiple tables/primary keys which means that the coordinator that gets the request has to go to multiple nodes to satisfy the node that holds each key. That puts a lot of pressure on the coordinator node that has to accept that request. It could end of actually having to hit all the nodes in your cluster (depending on batch and cluster size) and since it's Logged that log has to replicate n nodes as well. So it's a very expensive write in a moderate performance use case. 

A better approach with batching is to group only the writes that hit a single vertex(primary key) together if possible and make a batch out of that and single exec the others which will allow gocql to route the primary key to the appropriate node in the cluster if DiscoverHosts is set. It could be that  your batches are 2-5 writes instead of a whole block of random keys. 

I'm also assuming you're using batching for consistency across tables and not for performance reasons?